config/tools/man/ccs_tool.8
.TH "ccs_tool" "8" "" "" ""
.SH "NAME"
ccs_tool \- The tool used to make online queries to the cluster configuration.

.SH "SYNOPSIS"
.B ccs_tool
[\fIOPTION\fR].. <\fBcommand\fP>

.SH "DESCRIPTION"

\fBccs_tool\fP is part of the Cluster Configuration System (CCS). It used
to peform different kind of queries to the cluster configuration and has support
for some cluster.conf editing functions.

.SH "OPTIONS"
.TP 
\fB\-h\fP
Help.  Print out the usage.
.TP 
\fB\-V\fP
Print the version information.

sub\-commands have their own options, see below for more detail
.SH "COMMANDS"

.TP
\fBquery\fP \fI<xpath query>\fP
Perform an xpath query on running cluster configuration.

.TP 
\fBaddnode\fP [options] \fI<node> [<fenceoption=value>]...\fP
Adds a new node to the cluster configuration file. Fencing device options
are specified as key=value pairs (as many as required) and are entered into the
configuration file as is. See the documentation for your fencing agent for more
details (eg a powerswitch fence device may need to know which port the node is
connected to).
.br
\fIOptions:\fP
.br
\-v <votes>        Number of votes for this node (mandatory)
.br 
\-n <nodeid>       Node id for this node (optional)
.br
\-i <interface>    Network interface to use for this node. Mandatory if the cluster 
is using multicast as transport. Forbidden if not.
.br
\-m <multicast>    Multicast address for cluster. Only allowed on the first node to 
be added to the file. Subsequent nodes will use either multicast or broadcast 
depending on the properties of the first node.
.br
\-f <fencedevice>  Name of fence device to use for this node. The fence device 
section must already have been added to the file, probably using the addfence command.
.br
\-c <file>         Config file to use. Defaults to /etc/cluster/cluster.conf
.br
\-o <file>         Output file. Defaults to the same as -c



.TP 
\fBdelnode\fP [options] \fI<node>\fP
Delete a node from the cluster configuration file. Note: there is no 
"edit" command so to change the properties of a node you must delete it
and add it back in with the new properties.
.br
\fIOptions:\fP
.br
\-c <file>         Config file to use. Defaults to /etc/cluster/cluster.conf
.br
\-o <file>         Output file. Defaults to the same as -c



.TP 
\fBaddfence\fP [options] \fI<name> <agent> [<option>=<value>]...\fP
Adds a new fence device section to the cluster configuration file. <agent> is the
name of the fence agent that controls the device. the options following are entered
as key-value pairs. See the fence agent documentation for details about these. eg:
you may need to enter the IP address and username/password for a powerswitch fencing
device.
.br
\fIOptions:\fP
.br
\-c <file>         Config file to use. Defaults to /etc/cluster/cluster.conf
.br
\-o <file>         Output file. Defaults to the same as -c

.TP 
\fBdelfence\fP [options] \fI<node>\fP
Deletes a fencing device from the cluster configuration file.
delfence will allow you to remove a fence device that is in use by nodes.
This is to allow changes to be made, but be aware that it may produce an
invalid configuration file if you don't add it back in again.
.br
\fIOptions:\fP
.br
\-c <file>         Config file to use. Defaults to /etc/cluster/cluster.conf
.br
\-o <file>         Output file. Defaults to the same as -c


.TP 
\fBlsnode [options] \fP
List the nodes in the configuration file. This is (hopefully obviously) not
necessarily the same as the nodes currently in the cluster, but it should
be a superset.
.br
\fIOptions:\fP
.br
\-v                Verbose. Lists all the properties of the node, and the
node-specific properties of the fence device too.
.br
\-c <file>         Config file to use. Defaults to /etc/cluster/cluster.conf


.TP 
\fBlsfence [options] \fP
List all the fence devices in the cluster configuration file.
.br
\fIOptions:\fP
.br
\-v                Verbose. Lists all the properties of the fence device rather
than just the names and agents.
.br
\-c <file>         Config file to use. Defaults to /etc/cluster/cluster.conf


.TP 
\fBcreate [options] \fP \fI<clustername>\fP
Create a new, skeleton, configuration file. Note that "create" on its own will 
not create a valid configuration file. Fence agents and nodes will need to be 
added to it before handing it over to ccsd. The new configuration file will
have a version number of 1. Subsequent addnode/delnode/addfence/delfence operations
will increment the version number by 1 each time.
.br
\fIOptions:\fP
.br
.br
\-c <file>         Config file to create. Defaults to /etc/cluster/cluster.conf

.TP 
\fBaddnodeids\fP
Adds node ID numbers to all the nodes in cluster.conf. In RHEL4, node IDs were optional
and assigned by cman when a node joined the cluster. In RHEL5 they must be pre-assigned
in cluster.conf. This command will not change any node IDs that are already set in 
cluster.conf, it will simply add unique node ID numbers to nodes that do not already
have them.


.SH "SEE ALSO"
cluster.conf(5)


debian/changelog
redhat-cluster (3.1.8-1.2) unstable; urgency=medium

  [ gregor herrmann ]
  * Non-maintainer upload.
  * Fix "hardcodes /usr/lib/perl5":
    make debian/libccs-perl.install executable and use $Config{vendorarch} to
    get dynamic path
    (Closes: #752811)

 -- Damyan Ivanov <dmn@debian.org>  Mon, 04 Aug 2014 12:34:59 +0000

redhat-cluster (3.1.8-1.1) unstable; urgency=low

  * Non-maintainer upload.
  * Drop clvm Depends for redhat-cluster-suite binary package.
    Cluster (clvm) support was dropped in lvm2 source package by removing
    the clvm binary package. Drop the Depends also in redhat-cluster-suite.
    (Closes: #697870)

 -- Salvatore Bonaccorso <carnil@debian.org>  Thu, 10 Jan 2013 21:40:28 +0100

redhat-cluster (3.1.8-1) unstable; urgency=low

  [ Andres Rodriguez ]
  * New upstream release
  * Apply all the changes from Ubuntu.
  * Bump Standards Version.
  * debian/source/format: 3.0 (quilt).
  * Switch to dh 7.
    - debian/control: Bump debhelper build-dep.
    - debian/rules: Switch.

  [ Martin Loschwitz ]
  * Thanks a lot to Andres Rodrigues for all the great work he did on this!
  * Removed debian/control.modules.in and debian/upgrade.txt due to 
    being unnecessary now or hopelessly outdated
  * Add myself to Uploaders: in debian/control
  * Switch to dh 9.
 -- Martin Loschwitz <madkiss@debian.org>  Sun, 05 Aug 2012 08:13:00 +0000

redhat-cluster (3.0.12-3.1) unstable; urgency=low

  * Non-maintainer upload.
  * Fix pending l10n issues. Debconf translations:
    - Brazilian Portuguese (Flamarion Jorge).  Closes: #650048
    - Dutch; (Jeroen Schot).  Closes: #657007
    - Polish (Michał Kułach).  Closes: #663689

 -- Christian Perrier <bubulle@debian.org>  Wed, 21 Mar 2012 07:55:19 +0100

redhat-cluster (3.0.12-3) unstable; urgency=low

  * Add missing build dependency: openais-dev. (Closes: #640346)
  * Added danish translation, thanks to Joe Dalton <joedalton2@yahoo.dk>.
    (Closes: #628209)
  * cman: depend on telnet | telnet-client to allow arbitrary telnet 
    incarnations. (Closes: #626155)
  * gfs2-tools: fix symlinks. (Closes: #625965)
  * cman: confict with fence-agents. (Closes: #639985)

 -- Frederik Schüler <fs@debian.org>  Wed, 16 Nov 2011 18:57:50 +0100

redhat-cluster (3.0.12-2) unstable; urgency=low

  * Upload to unstable

 -- Guido Günther <agx@sigxcpu.org>  Fri, 02 Jul 2010 10:59:21 +0200

redhat-cluster (3.0.12-1) experimental; urgency=low

  * [a45231b] New patch 0005-Remove-bashisms-from-resource-scripts.patch
    Remove bashisms from resource scripts (Closes: #581137)
  * [3df4d28] Imported Upstream version 3.0.12
  * [25ffc05] Install rgmanager manpage
  * [7e5f87f] Update jp translation (Closes: #582722) - thanks to Hideki
    Yamane

 -- Guido Günther <agx@sigxcpu.org>  Mon, 14 Jun 2010 15:17:44 +0200

redhat-cluster (3.0.11-1) experimental; urgency=low

  * [9c55372] reformat ja.po
  * [ca106f6] Imported Upstream version 3.0.11
  * [03336a5] Grab libdlmcontrol.so.3.1
  * [c14c894] Enable pacemaker support - thanks a lot to Jörg Bachmann for
    testing
  * [287d065] New patch 0004-Depend-on-remote_fs.patch Depend on $remote_fs
  * [54bfc48] Bump (build-)dep on corosync.

 -- Guido Günther <agx@sigxcpu.org>  Wed, 05 May 2010 18:00:47 +0200

redhat-cluster (3.0.9-2) experimental; urgency=low

  * Upload to unstbale
  * [fb1da7e] Unignore po files ignored by upstreams .gitignore file
  * [6845591] Add new vi translation (Closes: #576145) - thanks to Clytie
    Siddall
  * [aa6bf55] Add jp debconf translation (Closes: #554841) - thanks to Hideki
    Yamane 

 -- Guido Günther <agx@sigxcpu.org>  Sat, 01 May 2010 10:47:32 +0200

redhat-cluster (3.0.9-1) experimental; urgency=low

  * [7c1a75c] Imported Upstream version 3.0.7
  * [9381ece] Drop 0003-Do-nothing-if-cman_tool-isn-t-there.patch fixed
    upstream
  * [8ed7908] Bump (build-)dep on corosync
  * [5b162fa] Enable gfs since it defaults to off now upstream.
  * [1d4ad8f] Imported Upstream version 3.0.9
  * [09c214b] Version dependencies of meta package
  * [4352b38] Redo patches
  * [61487cc] New patch 0003-Don-t-rely-on-chkconfig-and-check-for-
    network-manage.patch Don't rely on chkconfig and check for network-
    manager

 -- Guido Günther <agx@sigxcpu.org>  Wed, 10 Mar 2010 22:29:05 +0100

redhat-cluster (3.0.6-5) unstable; urgency=low

  * [ebda28c] Don't ship drbd resource script (Closes: #566095)

 -- Guido Günther <agx@sigxcpu.org>  Fri, 22 Jan 2010 20:45:52 +0100

redhat-cluster (3.0.6-4) unstable; urgency=low

  * [9725a65] Only install schema (Closes: #565382)

 -- Guido Günther <agx@sigxcpu.org>  Fri, 15 Jan 2010 16:39:40 +0100

redhat-cluster (3.0.6-3) unstable; urgency=low

  * Upload to unstable
  * [19bd29c] Depend on cman (Closes: #563787)
  * [2a019fa] Genereate proper depends for lcsos
  * [e276e1d] Add schema

 -- Guido Günther <agx@sigxcpu.org>  Mon, 11 Jan 2010 21:00:27 +0100

redhat-cluster (3.0.6-2) experimental; urgency=low

  * [e16890c] Update de.po (Closes: #562707) - thanks to Helge Kreutzmann
  * [e1b336d] Update pt debconf translation (Closes: #563015) - thanks to
    Américo Monteiro
  * [e01aa76] Update ru debconf translation (Closes: #563345) - thanks to Yuri
    Kozlov
  * [f959fd1] Update sv debconf translation (Closes: #562928) - thanks to
    Martin Bagge
  * [bbb4b4a] Run debconf-updatepo

 -- Guido Günther <agx@sigxcpu.org>  Sat, 02 Jan 2010 14:53:45 +0100

redhat-cluster (3.0.6-1) experimental; urgency=low

  * [b3b823a] Imported Upstream version 3.0.6
  * [3e69621] run debconf-updatepo during clean
  * [cb83708] build-depend on po-debconf
  * [b8b59d7] remove one more trailing whitespace
  * [ccd9a39] Depend on libxml2-utils for ccs_config_validate
  * [ff09490] redo patches
  * [08a23fa] Do nothing if cman_tool isn't there so we don't report an error
    if the package was removed but not purged.
  * [bbec984] Drop 0003-add-LD_LIBRARY_PATH-so-we-find-liblogthread.patch
    - applied upstream.

config/tools/ccs_tool/ccs_tool.c
#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <string.h>
#include <errno.h>

#include "copyright.cf"
#include "editconf.h"
#include "ccs.h"


/*
 * Old libccs retruned -error (mostly!) but didn't set errno (sigh)
 * New libccs sets errno correctly
 */
static char *errstring(int retcode)
{
	return strerror(errno);
}

static void tool_print_usage(FILE *stream);

int globalverbose=0;

static void test_print_usage(FILE *stream);

static int test_main(int argc, char *argv[], int old_format){
  int desc=0;
  int i=0;
  int error = 0;
  int force = 0, blocking = 0;
  char *str=NULL;
  char *cluster_name = NULL;

  if(argc <= 1){
    test_print_usage(stderr);
    exit(EXIT_FAILURE);
  }

  for(i=1; i < argc; i++){
    if(!strcmp(argv[i], "-h")){
      test_print_usage(stdout);
      exit(EXIT_SUCCESS);
    }
    if(!strcmp(argv[i], "-V")){
      printf("%s %s (built %s %s)\n", argv[0], RELEASE_VERSION, __DATE__, __TIME__);
      printf("%s\n", REDHAT_COPYRIGHT);
      exit(EXIT_SUCCESS);
    }
  }

  if(!strcmp(argv[1], "connect")){
    for(i=2; i < argc; i++){
      if(!strcmp(argv[i], "force")){
	printf("Force is set.\n");
	force = 1;
      } else if(!strcmp(argv[i], "block")){
	printf("Blocking is set.\n");
	blocking = 1;
      } else {
	cluster_name = argv[i];
	printf("Setting cluster name to %s\n", cluster_name);
      }
    }
    if(blocking && !force){
      fprintf(stderr, "Blocking can only be used with \"force\".\n");
      exit(EXIT_FAILURE);
    }
    if(force){
      desc = ccs_force_connect(cluster_name, blocking);
    } else {
      if(cluster_name){
	fprintf(stderr, "A cluster name can only be specified when using 'force'.\n");
	exit(EXIT_FAILURE);
      }
      desc = ccs_connect();
    }
    if(desc < 0){
      fprintf(stderr, "ccs_connect failed: %s\n", errstring(-desc));
      exit(EXIT_FAILURE);
    } else {
      printf("Connect successful.\n");
      printf(" Connection descriptor = %d\n", desc);
      ccs_disconnect(desc);
    }
  }
  else if(!strcmp(argv[1], "disconnect")){
    if(argc < 3){
      fprintf(stderr, "Wrong number of arguments.\n");
      exit(EXIT_FAILURE);
    }
    desc = ccs_connect();
    if((error = ccs_disconnect(desc))){
      fprintf(stderr, "ccs_disconnect failed: %s\n", errstring(-error));
      exit(EXIT_FAILURE);
    } else {
      printf("Disconnect successful.\n");
    }
  }
  else if(!strcmp(argv[1], "get")){
    if(argc < 4){
      fprintf(stderr, "Wrong number of arguments.\n");
      exit(EXIT_FAILURE);
    }
    desc = ccs_connect();
    if((desc < 0) || (error = ccs_get(desc, argv[3], &str))){
      fprintf(stderr, "ccs_get failed: %s\n", errstring(-error));
      exit(EXIT_FAILURE);
    } else {
	    if (old_format) {
		    printf("Get successful.\n");
		    printf(" Value = <%s>\n", str);
	    }
	    else {
		    printf("%s\n", str);
	    }
      if(str)free(str);
      ccs_disconnect(desc);
    }
  }
  else {
    fprintf(stderr, "Unknown command: %s\n", argv[1]);
    exit(EXIT_FAILURE);
  }

  exit(EXIT_SUCCESS);
}

static void test_print_usage(FILE *stream)
{
	fprintf(stream,
	  "Usage:\n"
	  "\n"
	  "ccs_test [Options] <Command>\n"
	  "\n"
	  "Options:\n"
	  "  -h                        Print usage.\n"
	  "  -V                        Print version information.\n"
	  "\n"
	  "Commands:\n"
	  "  connect <force> <block>   Connect to CCS and return connection descriptor.\n"
	  "  disconnect <desc>         Disconnect from CCS.\n"
	  "  get <desc> <request>      Get a value from CCS.\n"
	  );
}

static int xpath_query(int argc, char **argv)
{
	int handle;
	char *ret;
	int i;

	if (argc < 2) {
		fprintf(stderr,
			"Usage:\n"
			"\n"
			"ccs_tool query [-n] <xpath query>\n"
			"\n"
		        "options\n"
			"  -n     disable full XPath parsing\n");
		return 1;
	}

	/* Tell the library we want full XPath parsing */
	fullxpath = 1;

	if (strcmp(argv[1], "-n") == 0) {
		argv++;
		argc--;

		/* Actually ... no we don't */
		fullxpath = 0;
	}

	handle = ccs_connect();

	/* Process all the queries on the command-line */
	for (i=1; i<argc; i++) {
		if (!ccs_get(handle, argv[1], &ret)) {
			printf("%s\n", ret);
			free(ret);
		}
		else {
			fprintf(stderr, "Query failed: %s\n", strerror(errno));
			ccs_disconnect(handle);
			return -1;
		}
	}
	ccs_disconnect(handle);
	return 0;
}

static int tool_main(int argc, char *argv[])
{
  optind = 1;

  if (argc < 2 || !strcmp(argv[optind], "-h")) {
      tool_print_usage(stdout);
      exit(EXIT_SUCCESS);
  }
  if (!strcmp(argv[optind], "-V")) {
      printf("%s %s (built %s %s)\n", argv[0], RELEASE_VERSION,
	     __DATE__, __TIME__);
      printf("%s\n", REDHAT_COPYRIGHT);
      exit(EXIT_SUCCESS);
  }

  if(optind < argc){
    if(!strcmp(argv[optind], "-verbose")){
      optind++;
      globalverbose=1;
    }
    if(!strcmp(argv[optind], "help")){
      tool_print_usage(stdout);
      exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "query")){
	    return xpath_query(argc-1, argv+1);
    }
    else if(!strcmp(argv[optind], "addnode")){
	    add_node(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "delnode")){
	    del_node(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "addservice")){
	    add_service(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "delservice")){
	    del_node(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "addscript")){
	    add_script(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "delscript")){
	    del_node(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "addip")){
	    add_ip(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "delip")){
	    del_node(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "addfs")){
	    add_fs(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "delfs")){
	    del_node(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "addfdomain")){
	    add_fdomain(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "delfdomain")){
	    del_node(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "addfence")){
	    add_fence(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "delfence")){
	    del_fence(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "lsnode")){
	    list_nodes(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "lsservice")){
	    list_services(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "lsfence")){
	    list_fences(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "lsscript")){
	    list_scripts(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "lsip")){
	    list_ips(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "lsfs")){
	    list_fs(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "lsfdomains")){
	    list_fdomains(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "create")){
	    create_skeleton(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "addnodeids")){
	    add_nodeids(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }

    else {
      fprintf(stderr, "Unknown command, %s.\n"
	      "Try 'ccs_tool help' for help.\n", argv[optind]);
      exit(EXIT_FAILURE);
    }
  } else {
    fprintf(stderr, "Too few arguments.\n"
	    "Try 'ccs_tool help' for help.\n");
    exit(EXIT_FAILURE);
  }
  exit(EXIT_SUCCESS);
}

static void tool_print_usage(FILE *stream){
  fprintf(stream,
	  "Usage:\n"
	  "  ccs_tool [options] <command>\n"
	  "\n"
	  "Options:\n"
	  "  -verbose            Make some operations print more details.\n"
	  "  -h                  Print this usage and exit.\n"
	  "  -V                  Print version information and exit.\n"
	  "\n"
	  "Commands:\n"
	  "  help                Print this usage and exit.\n"
	  "  query <xpath query> Query the cluster configuration.\n"
	  "  addnode <node>      Add a node\n"
          "  delnode <node>      Delete a node\n"
          "  lsnode              List nodes\n"
	  "  addservice <name>   Add a service\n"
          "  delservice <name>   Delete a service\n"
          "  lsservice           List services\n"
          "  lsfence             List fence devices\n"
	  "  addfence <fencedev> Add a new fence device\n"
	  "  delfence <fencedev> Delete a fence device\n"
	  "  addscript <name>    Add a script resource\n"
	  "  delscript <name>    Delete a script resource\n"
	  "  lsscript            List script resources\n"
	  "  addip <name>        Add an IP address resource\n"
	  "  delip <name>        Delete an IP address resource\n"
	  "  lsip                List IP address resources\n"
	  "  addfs <name>        Add an filesystem resource\n"
	  "  delfs <name>        Delete an filesystem resource\n"
	  "  lsfs                List filesystem resources\n"
	  "  addfdomain <name>   Add an failover domain\n"
	  "  delfdomain <name>   Delete an failover domain\n"
	  "  lsfdomains          List failover domains\n"
	  "  create              Create a skeleton config file\n"
	  "  addnodeids          Assign node ID numbers to all nodes\n"
	  "\n");
}


int main(int argc, char *argv[])
{
	char *name = strdup(argv[0]);

	/*
	 * Don't be anal about the binary name.
	 * We expect either 'ccs_tool' or 'ccs_test',
	 * but interpret anything other than 'ccs_test'
	 * as 'ccs_tool'.
	 * That's not a bug, it's a feature.
	 */

	if (strcmp(basename(name), "ccs_test") == 0)
		return test_main(argc, argv, 1);
	else
		return tool_main(argc, argv);
}
